### DISCUSSION TITLE
USB Mass Storage Gadget

### DISCUSSION THREAD

**Discussion by probonopd** [rsta2/circle#495] [Ideas]
Circle has USB gadget mode, could it be made to support USB mass storage mode so that when attached to a host computer via USB, the contents of the microSD card could be accessed? This might be neat for embedded devices in which the microSD card is not easily accessible inside the case.

**Comment by rsta2** [comment]
This is already supported. See the test program in [test/usb-msd-gadget](https://github.com/rsta2/circle/tree/master/test/usb-msd-gadget). But as with the USB MIDI gadget the system must be configured to use the USB MSD gadget. It cannot be used at the same time with USB host mode or with the USB MIDI gadget.

**Reply by probonopd** [reply]
Thanks @rsta2, great to hear. I was hoping it could be enabled alongside USB MIDI gadget; is there a technical reason that this is not possible?

**Reply by rsta2** [reply]
At the moment only one gadget can be active at a time. This is a limitation in Circle.

**Reply by sebastienNEC** [reply]
Actually, I am working on a single composite MSD + CDC gadget. The same code could be adapted for a single MSD + MIDI gadget.

However, it doesn't quite work at the moment. Could @rsta2 have a look at my code (attached) and guess what's wrong ?

 I am basically deriving from CUSBMSDGadget and adding the CUSBCDCGadgetEndpoints on top. I have to make 2 modification on the USB core code:
* Making CUSBCompositeMSDCDCGadget a friend class of CUSBMSDGadget
* Changing the CUSBCDCGadgetEndpoints constructor to take in a CDWUSBGadget argument instead of a CUSBCDCGadget one.

If I comment out the CreateDevice + OnSuspend methods (and thus use the standard CUSBMSDGadget ones), the /dev/sdb1 disk is created.
With my methods though, circle seems to call CreateDevice twice without calling OnSuspend in between, which crashes on the assert.

Any idea why ?
Thank you !
[usbcomposite.zip](https://github.com/user-attachments/files/17952342/usbcomposite.zip)

**Reply by rsta2** [reply]
If `CreateDevice()` is called twice, it's likely, that the host sends the `SET_CONFIGURATION` request twice. That's not forbidden, but the Circle gadget code cannot handle this at the moment. Can you please try this patch, which simple ignores a second request:

```
diff --git a/lib/usb/gadget/dwusbgadget.cpp b/lib/usb/gadget/dwusbgadget.cpp
index c6a3c63..4732302 100644
--- a/lib/usb/gadget/dwusbgadget.cpp
+++ b/lib/usb/gadget/dwusbgadget.cpp
@@ -752,6 +752,11 @@ boolean CDWUSBGadget::SetConfiguration (u8 uchConfiguration)
 		return FALSE;
 	}
 
+	if (m_State == StateConfigured)
+	{
+		return TRUE;
+	}
+
 	assert (m_State == StateEnumDone);
 	m_State = StateConfigured;
 
```

**Reply by rsta2** [reply]
@sebastienNEC Did this help?

**Comment by sebastienNEC** [comment]
Hello Rene, sorry for not replying, because I have not gotten to the bottom of this yet. Yes it helped, in the sense that Circle does not crash anymore, but Linux is still not creating the CDC/MSD devices. I need to debug further... :)

**Reply by rsta2** [reply]
No problem. Thanks for info! I think I should add this patch to Circle.

**Reply by rsta2** [reply]
The patch has been added on the *develop* branch.
